src/query/api/v1/handler/graphite/find.go

```diff
@@ -57,52 +59,100 @@ func NewFindHandler(
 	}
 }
 
+func mergeTags(
+	terminatedResult *storage.CompleteTagsResult,
+	childResult *storage.CompleteTagsResult,
+) (map[string]bool, error) {
+	// sanity check the case.
+	if terminatedResult.CompleteNameOnly {
+		return nil, errors.New("terminated result is completing name only")
+	}
+
+	if childResult.CompleteNameOnly {
+		return nil, errors.New("child result is completing name only")
+	}
+
+	mapLength := len(terminatedResult.CompletedTags) + len(childResult.CompletedTags)
+	tagMap := make(map[string]bool, mapLength)
+
+	for _, tag := range terminatedResult.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = false
+		}
+	}
+
+	// NB: fine to overwrite any tags which were present in the `terminatedResult` map
+	// since if they appear in `childResult`, then they exist AND have children.
+	for _, tag := range childResult.CompletedTags {
+		for _, value := range tag.Values {
+			tagMap[string(value)] = true
+		}
+	}
+
+	return tagMap, nil
+}
+
 func (h *grahiteFindHandler) ServeHTTP(
 	w http.ResponseWriter,
 	r *http.Request,
 ) {
 	ctx := context.WithValue(r.Context(), handler.HeaderKey, r.Header)
 	logger := logging.WithContext(ctx)
 	w.Header().Set("Content-Type", "application/json")
-	query, rErr := parseFindParamsToQuery(r)
+
+	// NB: need to run two separate queries, one of which will match only the
+	// provided matchers, and one which will match the provided matchers with at
+	// least one more child node. For further information, refer to the comment
+	// for parseFindParamsToQueries
+	terminatedQuery, childQuery, raw, rErr := parseFindParamsToQueries(r)
 	if rErr != nil {
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
 		return
 	}
 
-	opts := storage.NewFetchOptions()
-	result, err := h.storage.FetchTags(ctx, query, opts)
-	if err != nil {
-		logger.Error("unable to complete tags", zap.Error(err))
-		xhttp.Error(w, err, http.StatusBadRequest)
-		return
-	}
+	var (
+		terminatedResult *storage.CompleteTagsResult
+		childResult      *storage.CompleteTagsResult
+		wg               sync.WaitGroup
+		multiErr         = xerrors.NewMultiError()
+		opts             = storage.NewFetchOptions()
+	)
 
-	partCount := graphite.CountMetricParts(query.Raw)
-	partName := graphite.TagName(partCount - 1)
-	seenMap := make(map[string]bool, len(result.Metrics))
-	for _, m := range result.Metrics {
-		tags := m.Tags.Tags
-		index := 0
-		// TODO: make this more performant by computing the index for the tag name.
-		for i, tag := range tags {
-			if bytes.Equal(partName, tag.Name) {
-				index = i
-				break
-			}
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		var err error
+		terminatedResult, err = h.storage.CompleteTags(ctx, terminatedQuery, opts)
+		if err != nil {
+			multiErr = multiErr.Add(err)
```

Comment:
nit: Need to do locking on this `multiErr`.

Instead perhaps just keep both separate then use a `xerrors.FirstError(terminatedErr, childErr)` instead of `multiErr.FinalError()`?

Reply:
Ah my bad, we have another multiErr type somewhere which is locking; I like the FirstError suggestion more anyway :+1: